scio-parquet/src/main/scala/com/spotify/scio/parquet/avro/ParquetAvroIO.scala

```diff
@@ -159,38 +169,102 @@ object ParquetAvroIO {
       // org.apache.beam.sdk.coders.AvroCoder.
       if (!isSpecific) {
-        hadoopConf.setBoolean(AvroReadSupport.AVRO_COMPATIBILITY, false)
-        hadoopConf.set(
+        conf.setBoolean(AvroReadSupport.AVRO_COMPATIBILITY, false)
+        conf.set(
           AvroReadSupport.AVRO_DATA_SUPPLIER,
           classOf[GenericDataSupplier].getCanonicalName
         )
       }
 
-      AvroReadSupport.setAvroReadSchema(hadoopConf, readSchema)
+      AvroReadSupport.setAvroReadSchema(conf, readSchema)
       if (projection != null) {
-        AvroReadSupport.setRequestedProjection(hadoopConf, projection)
+        AvroReadSupport.setRequestedProjection(conf, projection)
       }
       if (predicate != null) {
-        ParquetInputFormat.setFilterPredicate(hadoopConf, predicate)
+        ParquetInputFormat.setFilterPredicate(conf, predicate)
       }
 
-      val tCoder = sc.pipeline.getCoderRegistry.getCoder(ScioUtil.classOf[T])
+      val tCoder = CoderMaterializer.beam(sc, coder)
       val cleanedProjectionFn = ClosureCleaner.clean(projectionFn)
 
       sc.applyTransform(
         ParquetRead.read[A, T](
           ReadSupportFactory.avro,
-          new SerializableConfiguration(hadoopConf),
+          new SerializableConfiguration(conf),
           path,
           Functions.serializableFn(cleanedProjectionFn)
         )
       ).setCoder(tCoder)
     }
+
+    private def readLegacy(sc: ScioContext, conf: Configuration, path: String)(implicit
+      coder: Coder[T]
+    ): SCollection[T] = {
+      val job = Job.getInstance(conf)
+      GcsConnectorUtil.setInputPaths(sc, job, path)
+      job.setInputFormatClass(classOf[AvroParquetInputFormat[T]])
+      job.getConfiguration.setClass("key.class", classOf[Void], classOf[Void])
+      job.getConfiguration.setClass("value.class", avroClass, avroClass)
+
+      // Needed to make GenericRecord read by parquet-avro work with Beam's
+      // org.apache.beam.sdk.coders.AvroCoder.
+      if (ScioUtil.classOf[A] == classOf[GenericRecord]) {
+        job.getConfiguration.setBoolean("parquet.avro.compatible", false)
+        job.getConfiguration.set(
+          "parquet.avro.data.supplier",
+          "org.apache.parquet.avro.GenericDataSupplier"
+        )
+      }
+
+      AvroParquetInputFormat.setAvroReadSchema(job, readSchema)
+      if (projection != null) {
+        AvroParquetInputFormat.setRequestedProjection(job, projection)
+      }
+      if (predicate != null) {
+        ParquetInputFormat.setFilterPredicate(job.getConfiguration, predicate)
+      }
+
+      val g = ClosureCleaner.clean(projectionFn) // defeat closure
+      val aCls = avroClass
+      val oCls = ScioUtil.classOf[T]
+      val transform = HadoopFormatIO
+        .read[JBoolean, T]()
+        // Hadoop input always emit key-value, and `Void` causes NPE in Beam coder
+        .withKeyTranslation(new SimpleFunction[Void, JBoolean]() {
+          override def apply(input: Void): JBoolean = true
+        })
```

Comment:
Can't we write ?
```suggestion
        .withKeyTranslation((input: Void) => true: JBoolean)
```

Reply:
sadly no, we need to fully write it out for Scala 2.12 interop 😭

Reply:
Yes, `SimpleFunction` is not an interface but we have an helper for that
```suggestion
        .withKeyTranslation(Functions.simpleFn[Void, JBoolean](_ => true))
```